File: Linked-List-I/050-insert-node/insert-node.js
class Node {
  constructor(val) {
    this.val = val;
    this.next = null;
  }
}

// O(n) time and O(1) space where n is the number of nodes
const insertNode = (head, value, index) => {
  if (index === 0) {
    const newNode = new Node(value);
    newNode.next = head;
    return newNode;
  }
  let current = head;
  let count = 0; 

  while (current !== null) {
    if (count === index - 1) {
      const newNode = new Node(value);
      const tempNext = current.next;
      current.next = newNode;
      newNode.next = tempNext;
      return head;
    }
    count += 1;
    current = current.next;
  }
};

module.exports = {
  insertNode,
};

// Write a function, insertNode, that takes in the head of a linked list, a value, and an index. The function should insert a new node with the value into the list at the specified index. Consider the head of the linked list as index 0. The function should return the head of the resulting linked list.

// Do this in-place.

// You may assume that the input list is non-empty and the index is not greater than the length of the input list.

//  x -> a  ->  b  -> c 
// curr

// count = 1
// insertNode(a, 'x', 0);

File: Linked-List-I/054-insert-node/insert-node.py
class Node:
  def __init__(self, val):
    self.val = val
    self.next = None

# O(n) time and O(1) space
def insert_node(head, value, index):
  if index == 0:
    new_node = Node(value)
    new_node.next = node
    return new_node

  current = head
  prev = None

  while current is not None:
    if index == 0:
      new_node = Node(value)
      new_node.next = current
      prev.next = new_node
      return head

    prev = current
    current = current.next
    index -= 1

  prev.next = Node(value)
  return head

# insert_node(a, x, 2)
# a -> b -> c
# current = Node(c)
# prev = Node(b)
# index = 0
# new_head = Node(x)
